### PR TITLE
Move the type traits submodule to utility

### PR DESCRIPTION
**Pull request by joergi-w**
_For the reviewer:_ I strongly recommend to **review commit-wise**! I created separate commits for moving the files and creating the deprecation note, so the moved file contents do not appear (you see ~90% less). 

This PR moves the files from `core/type_traits` to the position specified below (in the deleted paths there are still header files with a deprecation note). 

`all.hpp basic.hpp concept.hpp pre.hpp` ⇒ `utility/type_traits/`
`transformation_trait_or.hpp` ⇒ `utility/type_traits/detail/`
`function.hpp` ⇒ `utility/type_traits/function_traits.hpp` and `utility/detail/multi_invocable.hpp`
`lazy.hpp` ⇒ `utility/type_traits/lazy_conditional.hpp` and `core/detail/is_class_template_declarable.hpp`
`deferred_crtp_base.hpp template_inspection.hpp` ⇒ `core/detail/`
`iterator.hpp` ⇒ `core/detail/iterator_traits.hpp`
`range.hpp` ⇒ `core/range/type_traits.hpp` (remove range_compatible)
`pack.hpp` ⇒ deleted

part of https://github.com/seqan/product_backlog/issues/160

### CODE REVIEW

**Comment by smehringer** [comment]
Please use the `[MISC]` tag for the commits :)

**Comment by joergi-w** [comment]
> Please use the `[MISC]` tag for the commits :)

Thanks! :)

File: include/seqan3/utility/type_traits/all.hpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides various type traits.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+/*!\defgroup type_traits Type Traits
+ * \brief Provides various type traits and their shortcuts.
+ * \ingroup core
```
Comment (eseiler) [inline include/seqan3/utility/type_traits/all.hpp:15]:
```suggestion
 * \ingroup utility
```

**Review by eseiler — CHANGES_REQUESTED** [review]
Can you re-add the `#pragma once` in the deprecated headers? You sometimes removed it.

**Comment by joergi-w** [comment]
> Can you re-add the `#pragma once` in the deprecated headers? You sometimes removed it.

Sure! I also created all.hpp for Utility module, in order to define the `utility` group for doxygen.

**Review by eseiler — APPROVED** [review]
Thank you

File: include/seqan3/core/range/type_traits.hpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+/*!\addtogroup type_traits
+ * \{
+ */
+
+// ----------------------------------------------------------------------------
+// value_type
+// ----------------------------------------------------------------------------
+
+#ifdef SEQAN3_DEPRECATED_310
```
Comment (smehringer) [inline include/seqan3/core/range/type_traits.hpp:64]: why did you deprecate all of this? I thought only range_compatible should be deprecated?

This would have been easier if it would have been done in a separate PR ;) for next time.

File: include/seqan3/core/type_traits/all.hpp [hunk 3]
```diff
@@ -6,21 +6,18 @@
...
-#include <seqan3/core/type_traits/template_inspection.hpp>
-#include <seqan3/core/type_traits/transformation_trait_or.hpp>
+#include <seqan3/core/detail/deferred_crtp_base.hpp>
+#include <seqan3/core/detail/iterator_traits.hpp>
+#include <seqan3/core/detail/template_inspection.hpp>
+#include <seqan3/core/range/type_traits.hpp>
+#include <seqan3/utility/type_traits/all.hpp>
+
+SEQAN3_DEPRECATED_HEADER("This header is deprecated and will be removed in SeqAn-3.1. "
+                         "Please #include <seqan3/utility/type_traits/all.hpp> instead.")
```
Comment (smehringer) [inline include/seqan3/core/type_traits/all.hpp:23]: Please leave the one liner even though it exceeds the line length. If you break the line here, unfortunately the second line is not properly displayed to the user if triggered.

Same for all other files

**Review by smehringer — CHANGES_REQUESTED** [review]
First batch :) I will review the rest tomorrow

File: include/seqan3/core/range/type_traits.hpp (hunk 2 above)
Comment (joergi-w) [reply include/seqan3/core/range/type_traits.hpp:64]: These `#ifdef SEQAN3_DEPRECATED_310` were already in the `type_traits/range.hpp` before and I didn't decide to deprecate this. I simply moved the whole file without changing its content (except removing range_compatible in the following commit). If you look at the individual commits it should become clearer where I moved a file or where I changed content.

File: include/seqan3/core/type_traits/all.hpp (hunk 3 above)
Comment (joergi-w) [reply include/seqan3/core/type_traits/all.hpp:23]: Oops, I didn't know... I'll change it!

File: include/seqan3/core/range/type_traits.hpp (hunk 2 above)
Comment (smehringer) [reply include/seqan3/core/range/type_traits.hpp:64]: Ah ok! sorry I missed that :)

**Comment by smehringer** [comment]
Oh oh.. a lot of merge conflicts. hopefully just concering the inlcudes. I hope its fine if I review after you resolved them. :)

**Comment by joergi-w** [comment]
Rebased on master.

**Comment by joergi-w** [comment]
Rebased on master.

**Review by smehringer — APPROVED** [review]
